jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/RequestListeners.java

```diff
@@ -0,0 +1,271 @@
+    {
+        try
+        {
+            if (listener != null)
+                listener.onCommit(request);
+        }
+        catch (Throwable x)
+        {
+            LOG.info("Exception while notifying listener {}", listener, x);
+        }
+    }
+
+    public void addContentListener(Request.ContentListener listener)
+    {
+        Request.ContentListener existing = contentListener;
+        contentListener = existing == null ? listener : (request, byteBuffer) ->
+        {
+            notifyContent(existing, request, byteBuffer);
+            notifyContent(listener, request, byteBuffer);
+        };
+    }
+
+    private static void notifyContent(Request.ContentListener listener, Request request, ByteBuffer byteBuffer)
+    {
+        try
+        {
+            if (listener != null)
+            {
+                // The same ByteBuffer instance may be passed to multiple listeners
+                // that may modify the position and limit, so clear it every time.
+                byteBuffer.clear();
+                listener.onContent(request, byteBuffer);
+            }
+        }
+        catch (Throwable x)
+        {
+            LOG.info("Exception while notifying listener {}", listener, x);
+        }
+    }
+
+    public void addSuccessListener(Request.SuccessListener listener)
+    {
+        Request.SuccessListener existing = successListener;
+        successListener = existing == null ? listener : request ->
+        {
+            notifySuccess(existing, request);
+            notifySuccess(listener, request);
+        };
+    }
+
+    private static void notifySuccess(Request.SuccessListener listener, Request request)
+    {
+        try
+        {
+            if (listener != null)
+                listener.onSuccess(request);
+        }
+        catch (Throwable x)
+        {
+            LOG.info("Exception while notifying listener {}", listener, x);
+        }
+    }
+
+    public void addFailureListener(Request.FailureListener listener)
+    {
+        Request.FailureListener existing = failureListener;
+        failureListener = existing == null ? listener : (request, failure) ->
+        {
+            notifyFailure(existing, request, failure);
+            notifyFailure(listener, request, failure);
+        };
+    }
+
+    private static void notifyFailure(Request.FailureListener listener, Request request, Throwable failure)
+    {
+        try
+        {
+            if (listener != null)
+                listener.onFailure(request, failure);
+        }
+        catch (Throwable x)
+        {
+            LOG.info("Exception while notifying listener {}", listener, x);
+        }
+    }
+
+    public void clear()
+    {
+        queuedListener = null;
+        beginListener = null;
+        headersListener = null;
+        commitListener = null;
+        contentListener = null;
+        successListener = null;
+        failureListener = null;
+    }
+
+    public static class Internal extends RequestListeners
```

Comment:
Internal but public? Something doesn't feel right about this pattern, too much internal stuff is exposed.